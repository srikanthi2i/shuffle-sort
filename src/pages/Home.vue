<template>
  <div class="home">
    <aside>
      <div class="action-container">
        <Button label="Shuffle" @shuffle-event="shuffle" eventAction="shuffleEvent" />
        <Button label="Sort" @sort-event="sort" eventAction="sortEvent" />
      </div>
    </aside>
    <div class="panel">
      <div class="card-container">
        <Card
          v-for="card in cards"
          :key="card"
          :value="card"
          :bg-color="'color-var--' + getRandomColorCode()"
        />
      </div>
    </div>
  </div>
</template>

<script lang="ts">
import { defineComponent } from 'vue';
import Button from '../components/Button.vue';
import Card from '../components/Card.vue';

export default defineComponent({
  name: 'Home',
  components: {
    Button,
    Card
  },
  data() {
    return {
      cards: [] as number[],
      cardLength: 9
    };
  },
  methods: {
    initializeCards() {
      let i = 1;
      while (i <= this.cardLength) {
        this.cards.push(i);
        i++;
      }
    },
    getRandomColorCode(): number {
      return (Math.floor(Math.random() * 10) % 4) + 1;
    },
    shuffle(): void {
      if (this.cardLength < 100) {
        this.cards.sort(() => Math.random() - 0.5);
      } else {
        let currentIndex = this.cardLength;
        let temporaryValue;
        let randomIndex;
        while (0 !== currentIndex) {
          randomIndex = Math.floor(Math.random() * currentIndex);
          currentIndex -= 1;
          temporaryValue = this.cards[currentIndex];
          this.cards[currentIndex] = this.cards[randomIndex];
          this.cards[randomIndex] = temporaryValue;
        }
      }
    },
    sort(): void {
      this.cards.sort();
    }
  },
  mounted() {
    this.initializeCards();
  }
});
</script>

<style lang="scss" scoped>
.home {
  display: flex;
  flex-wrap: wrap;
  padding: 0 2rem;
  aside {
    width: 100%;
    .action-container {
      display: flex;
      flex-wrap: wrap;
      justify-content: center;
    }
  }
  .panel {
    width: 100%;
    .card-container {
      display: flex;
      flex-wrap: wrap;
    }
  }
  @include media-desktop {
    padding: 0;
    .panel {
      width: 75%;
      order: 1;
    }
    aside {
      width: 25%;
      order: 2;
    }
  }
}
</style>
