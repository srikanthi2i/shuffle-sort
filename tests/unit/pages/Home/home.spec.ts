import { mount, MountingOptions, VueWrapper } from '@vue/test-utils';
import Home from '@/pages/Home.vue';

describe('Home.vue', () => {
  type Instance = InstanceType<typeof Home>;
  let mountFunction: (options?: MountingOptions<Instance>) => VueWrapper<Instance>;
  const cardLength = 9;

  beforeEach(() => {
    mountFunction = (options = {}) => {
      return mount(Home, {
        ...options
      });
    };
  });
  it('should render component and match snapshot', () => {
    const wrapper = mountFunction();
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should render cards after initialize', async () => {
    const initializeCardsSpy = jest.spyOn(Home.methods, 'initializeCards');
    const wrapper = mountFunction();
    expect(initializeCardsSpy).toHaveBeenCalled();
    await wrapper.setData({ cardLength, cards: [] });
    expect(wrapper.findAll('.card').length).toBe(cardLength);
  });

  it('should trigger the shuffle event', async () => {
    const wrapper = mountFunction();
    const shuffleButton = wrapper.findAll('button')[0];
    await wrapper.setData({ cardLength, cards: [] });
    expect(shuffleButton.text()).toBe('Shuffle');
    await wrapper.vm.$emit('shuffleEvent');
    expect(wrapper.emitted().shuffleEvent).toBeTruthy();
    await wrapper.vm.shuffle();
    expect(wrapper.vm.cards).not.toEqual([...wrapper.vm.cards].sort());
  });

  it('should trigger the sort event', async () => {
    const wrapper = mountFunction();
    const sortButton = wrapper.findAll('button')[1];
    await wrapper.setData({ cardLength, cards: [] });
    expect(sortButton.text()).toBe('Sort');
    await wrapper.vm.$emit('sortEvent');
    expect(wrapper.emitted().sortEvent).toBeTruthy();
    await wrapper.vm.sort();
    expect(wrapper.vm.cards).toEqual(wrapper.vm.cards.sort());
  });
});
