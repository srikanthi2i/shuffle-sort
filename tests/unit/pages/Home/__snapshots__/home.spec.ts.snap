// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home.vue should render component and match snapshot 1`] = `
<div class="home">
  <aside>
    <div class="action-container">
      <div class="button-wrapper"><button>Shuffle</button></div>
      <div class="button-wrapper"><button>Sort</button></div>
    </div>
  </aside>
  <div class="panel">
    <div class="card-container"></div>
  </div>
</div>
`;
